src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const newState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          newState[key] = action.extraData[key];
+        }
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      case 'clear':
+        for (const key in newState) {
+          delete newState[key];
+        }
+        break;
+      default:
+        break;
+    }
+    stateHistory.push({ ...newState });
```

Comment:
Instead of pushing the state copy inside the loop, which is an action performed in all cases, you should do it just once after the 'switch' statement to adhere to the 'Don't Repeat Yourself' principle.